nixos/lib/test-driver/test-driver.py

```diff
@@ -0,0 +1,680 @@
+#! /somewhere/python3
+
+from __future__ import print_function
+from xml.sax.saxutils import XMLGenerator
+import atexit
+import _thread
+import os
+import pty
+import queue
+import re
+import shutil
+import socket
+import subprocess
+import sys
+import tempfile
+import time
+import unicodedata
+
+CHAR_TO_KEY = {
+    'A': 'shift-a', 'N': 'shift-n',  '-': '0x0C', '_': 'shift-0x0C',
+    'B': 'shift-b', 'O': 'shift-o',  '=': '0x0D', '+': 'shift-0x0D',
+    'C': 'shift-c', 'P': 'shift-p',  '[': '0x1A', '{': 'shift-0x1A',
+    'D': 'shift-d', 'Q': 'shift-q',  ']': '0x1B', '}': 'shift-0x1B',
+    'E': 'shift-e', 'R': 'shift-r',  ';': '0x27', ':': 'shift-0x27',
+    'F': 'shift-f', 'S': 'shift-s',  "'": '0x28', '"': 'shift-0x28',
+    'G': 'shift-g', 'T': 'shift-t',  '`': '0x29', '~': 'shift-0x29',
+    'H': 'shift-h', 'U': 'shift-u', '\\': '0x2B', '|': 'shift-0x2B',
+    'I': 'shift-i', 'V': 'shift-v',  ',': '0x33', '<': 'shift-0x33',
+    'J': 'shift-j', 'W': 'shift-w',  '.': '0x34', '>': 'shift-0x34',
+    'K': 'shift-k', 'X': 'shift-x',  '/': '0x35', '?': 'shift-0x35',
+    'L': 'shift-l', 'Y': 'shift-y',  ' ': 'spc',
+    'M': 'shift-m', 'Z': 'shift-z', '\n': 'ret',
+    '!': 'shift-0x02', '@': 'shift-0x03', '#': 'shift-0x04', '$': 'shift-0x05',
+    '%': 'shift-0x06', '^': 'shift-0x07', '&': 'shift-0x08', '*': 'shift-0x09',
+    '(': 'shift-0x0A', ')': 'shift-0x0B',
+    }
+
+
+def eprint(*args, **kwargs):
+    print(*args, file=sys.stderr, **kwargs)
+
+
+def create_vlan(vlan_nr):
+    global log
+    log.log('starting VDE switch for network {}'.format(vlan_nr))
+    vde_socket = os.path.abspath('./vde{}.ctl'.format(vlan_nr))
+    pty_master, pty_slave = pty.openpty()
+    vde_process = subprocess.Popen(
+        'vde_switch -s {} --dirmode 0777'.format(vde_socket),
+        bufsize=0,  # decide what buffering is appropriate
+        stdin=pty_slave,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+        shell=True,
+        )
+    fd = os.fdopen(pty_master, 'w')
+    fd.write('version\n')
+    # TODO: perl version checks if this can be read from
+    # an if not, dies. we could hang here forever. Fix it.
+    vde_process.stdout.readline()
+    if not os.path.exists(os.path.join(vde_socket, 'ctl')):
+        raise Exception('cannot start vde_switch')
+
+    return (vlan_nr, vde_socket, vde_process, fd)
+
+
+def retry(fn):
+    """Call the given function repeatedly, with 1 second intervals,
+    until it returns True or a timeout is reached.
+    """
+
+    for n in range(900):
+        if fn(False):
+            return
+        time.sleep(1)
+
+    if not fn(True):
+        raise Exception('action timed out')
+
+
+class Logger:
```

Comment:
can we make use of python's native logging, by any chance?

This would avoid mixing the creating of log records with the different representations.

I assume nesting might be hard, but still wanted to ask :-)